torch/csrc/jit/codegen/cuda/lower_index_hoist.h

```diff
@@ -2,147 +2,90 @@
-  std::pair<Val*, bool> insert(
-      IterDomain* indexed_consumer_id,
-      TensorDomain* consumer_td,
-      const std::vector<IterDomain*>& loop_domains,
-      const std::unordered_map<IterDomain*, Val*>& loop_index_map,
-      const std::vector<kir::ForLoop*>& loops,
-      Val* index);
-
-  const auto& commonIndexMap() const {
-    return common_index_map_;
-  }
-
-  const auto& useCounts() const {
-    return use_counts_;
-  }
+  //! For the given index, insert its subexpressions to the loop that has
+  //! minimum amount of computation. For example, if I have a loop
+  //! FOR i1
+  //!   FOR i2
+  //!     FOR i3
+  //!       FOR i4
+  //!         index = ((i1*1 + i2*2) + i3*3) + i4*4
+  //! then this function will try insert i1*1 to common_index_map_[FOR i1],
+  //! try insert i1*1 + i2*2 to common_index_map_[FOR i2],
+  //! try insert ((i1*1 + i2*2) + i3*3) to common_index_map_[FOR i3],
+  //! try insert ((i1*1 + i2*2) + i3*3) + i4*4 to common_index_map_[FOR i4],
+  //! Before insertion, this function recursively uses
+  //! eliminateCommonSubexpression to find existing expressions/subexpressions
+  //! in common_index_map_ that can be reused. If a reuse oppportunity is found,
+  //! then this function will modify the definition of `index` to use the
+  //! existing subexpression. This function returns the modified value whose
+  //! definition reuses other expressions in the list.
+  Val* hoistIndex(Val* index, const std::vector<kir::ForLoop*>& loops);
+
+  //! common_index_map_ stores all seen indices in a given loop, however, we
+  //! don't want to create separate allocation for all of them. We are only
+  //! interested in allocating for the indices that is actually hoisted, or used
+  //! more than once. This method returns the Vals that will get its separate
+  //! allocation.
+  std::vector<Val*> getHoistedIndices(kir::ForLoop* loop) const;
 
  private:
-  //! Utility method to insert a key into common index
-  //!  map. Returns a pair of an IR node and a boolean value.
-  //! The IR node will be the previously inserted index if
-  //!  the key found a match, or will be the original index
-  //!  if this is new key and the key will be stored.
-  //! The boolean value will be true if the key is stored,
-  //!  i.e. first time it is inserted.
-  std::pair<Val*, bool> tryInsertNewIndex(CommonIndexKey key, Val* index);
+  //! This is the underlying implementation of the public hoistIndex, with some
+  //! additional arguments and return values.
+  //! Returns (hoisted index, has tensor dependency)
+  std::pair<Val*, bool> hoistIndex(
+      Val* index,
+      const std::vector<kir::ForLoop*>& loops,
+      int64_t position, // if `index` is given (i.e., is_give == true), then
+                        // this is the position of the outer-most loop nest that
+                        // contains all the dependencies of `index`. if `index`
+                        // is a subexpression of the given index, then this is
```

Comment:
A bit confused here: `index` is a subexpression of the given index. Which index are you referring to by "the given index"?

Reply:
"the given index" refers to the index given to the public method `hoistIndex`. I have renamed this function to `hoistIndexImpl` and updated the comment here to be more clear.